vpr/src/route/heap_type.h

```diff
@@ -159,4 +160,66 @@ class HeapInterface {
     virtual void free_all_memory() = 0;
 };
 
+template<typename T, typename RouteInf>
+void push_back_node(
+    T* heap,
+    const RouteInf& rr_node_route_inf,
+    int inode,
+    float total_cost,
+    int prev_node,
+    int prev_edge,
+    float backward_path_cost,
+    float R_upstream) {
+    /* Puts an rr_node on the heap with the same condition as node_to_heap,
+     * but do not fix heap property yet as that is more efficiently done from
+     * bottom up with build_heap    */
+
+    if (total_cost >= rr_node_route_inf[inode].path_cost)
+        return;
+
+    t_heap* hptr = heap->alloc();
+    hptr->index = inode;
+    hptr->cost = total_cost;
+    hptr->u.prev.node = prev_node;
+    hptr->u.prev.edge = prev_edge;
+    hptr->backward_path_cost = backward_path_cost;
+    hptr->R_upstream = R_upstream;
+
+    heap->push_back(hptr);
+}
+
+template<typename T, typename RouteInf>
+void add_node_to_heap(
+    T* heap,
+    const RouteInf& rr_node_route_inf,
+    int inode,
+    float total_cost,
+    int prev_node,
+    int prev_edge,
+    float backward_path_cost,
+    float R_upstream) {
+    /* Puts an rr_node on the heap, if the new cost given is lower than the     *
+     * current path_cost to this channel segment.  The index of its predecessor *
+     * is stored to make traceback easy.  The index of the edge used to get     *
+     * from its predecessor to it is also stored to make timing analysis, etc.  *
+     * easy.  The backward_path_cost and R_upstream values are used only by the *
+     * timing-driven router -- the breadth-first router ignores them.           */
+
+    if (total_cost >= rr_node_route_inf[inode].path_cost)
+        return;
+
+    t_heap* hptr = heap->alloc();
+
+    hptr->index = inode;
+    hptr->cost = total_cost;
+    VTR_ASSERT_SAFE(hptr->u.prev.node == NO_PREVIOUS);
+    VTR_ASSERT_SAFE(hptr->u.prev.edge == NO_PREVIOUS);
+    hptr->u.prev.node = prev_node;
+    hptr->u.prev.edge = prev_edge;
+    hptr->backward_path_cost = backward_path_cost;
+    hptr->R_upstream = R_upstream;
+
+    heap->add_to_heap(hptr);
+}
```

Comment:
I'd tend to classify these as router (rather than heap type) constructs. So they're probably better in router files (perhaps route_common.h or route_timing.h).

Additionally, I think the only difference between these two operations is whether we call add_to_heap() or push_back() at the end, so cost check and hptr setup should really be factored out in to a common function.

Reply:
Agreed, done